API Change: Stream Flash: Deprecating stream_flash_erase_page

de-nordic opened the pull request:
A few commits that
 - move stream flash logic to stop using stream_flash_erase_page
 - fixing tests
 - marking  the stream_flash_erase_page as deprecated

Fixes #67407

de-nordic committed:
storage/stream_flash: Switch to stream_flash_erase_to_append

Switches flash_sync to use stream_flash_erase_to_append
instead of stream_flash_erase_page in preparation of deprecating
the later and reming control on erase from API user.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
tests/stream_flash: Prepare for stream_flash_erase_page removal

Keep on testing the function but fix logic where applicable
and remove last_erased_page_start_offset which has been replaced
with erased_up_to in stream_flash_ctx.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
storage/stream_flash: Deprecate stream_flash_erase_page

The commit marks stream_flash_erase_page as deprecated.
After the function is removed users will have to rely on stream
flash to erase space it needs prior to write.

The stream_flash_erase_page has several issues like using non-relative
offset for erases, allowing to destroy data already written or even
erase out of area available for stream flash.

This fixes #67407

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nordicjm commented on subsys/storage/stream/Kconfig:
```diff
@@ -18,6 +18,22 @@ config STREAM_FLASH_ERASE
+	depends on STREAM_FLASH_ERASE
+	depends on FLASH_HAS_NO_EXPLICIT_ERASE
+	help
+	  This is used in stream flash code to exclude invocation of erase
```
n.b. for future: column length is 100 characters not 80

nordicjm commented on subsys/storage/stream/stream_flash.c:
```diff
@@ -72,7 +72,65 @@ static int settings_direct_loader(const char *key, size_t len,
+
+#if defined(CONFIG_STREAM_FLASH_ERASE_ONLY_WHEN_SUPPORTED)
+	/* There are both types of devices */
+	const struct flash_parameters *fparams = flash_get_parameters(ctx->fdev);
```
should declare variables at top

nordicjm commented on subsys/storage/stream/stream_flash.c:
```diff
@@ -72,7 +72,65 @@ static int settings_direct_loader(const char *key, size_t len,
+	} else {
+		ctx->erased_up_to += page.size;
+	}
+
```
```suggestion
```

nordicjm commented on subsys/storage/stream/stream_flash.c:
```diff
@@ -85,6 +143,11 @@ int stream_flash_erase_page(struct stream_flash_ctx *ctx, off_t off)
 		return -ERANGE;
 	}
 
+	/* Do not allow pages that has already been erased */
```
-has+have

nordicjm reviewed (dismissed):
minor nit.

kartben reviewed (changes requested):
- deprecated API should be mentioned in the release notes (with brief mention of what's the suggested replacement)
- doxygen comment of the deprecated API should indicate what's the suggested replacement so that people who will potentially start seeing warnings in their code know what to do
- https://github.com/zephyrproject-rtos/zephyr/issues/80539 to be updated with a bullet point referencing this PR so that future-you knows when it's time to actually drop the API.

thanks!

de-nordic commented:
> * deprecated API should be mentioned in the release notes (with brief mention of what's the suggested replacement)

Migration guide separately or in the deprecation note?

> 
>     * doxygen comment of the deprecated API should indicate what's the suggested replacement so that people who will potentially start seeing warnings in their code know what to do
> 
>     * [Deprecated APIs/options to be removed by 4.3 #80539](https://github.com/zephyrproject-rtos/zephyr/issues/80539) to be updated with a bullet point referencing this PR so that future-you knows when it's time to actually drop the API.
> 
> 
> thanks!

Sure, will do that. Note though that Stream Flash API is not considered stable, that is why deprecation process was not followed in the first place.

de-nordic committed:
doc: release-notes-4.1: stream_flash_erase_page gets deprecated

Add note on deprecating stream_flash_erase_page and how to substitute
its functionality.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

maass-hamburg commented:
> * deprecated API should be mentioned in the release notes (with brief mention of what's the suggested replacement)

@kartben maybe the [docs](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html) should be changed, so that entries in migration changes are also needed for experiential and unstable api, to document the established practice here.

de-nordic commented:
@kartben Can you help here? This
https://github.com/zephyrproject-rtos/zephyr/actions/runs/12930306331/job/36061561448?pr=82326
![image](https://github.com/user-attachments/assets/ebf0463e-3779-4dcd-8684-083103b025d0)
ain't caused by my lines? 
![image](https://github.com/user-attachments/assets/6a315b84-9399-45da-ae4c-ae26d04a2c1e)

kartben commented on doc/releases/release-notes-4.1.rst:
```diff
@@ -43,6 +43,14 @@ Deprecated in this release
+  can be achieved using :c:func:`flash_area_erase` or :c:func:`flash_erase`. Nevertheless
+  erasing of a device, while stream flash is supposed to do so, as configured, will result in
+  data lost from stream flash. There are only two situations where device should be erased
+  directly:
+    1) when Stream Flash is not configured to do erase on its own
+    2) when erase is used for removal of a data prior or after Stream Flash uses the designated area.
```
```suggestion
  directly:

  1. when Stream Flash is not configured to do erase on its own
  2. when erase is used for removal of a data prior or after Stream Flash uses the designated area.```

de-nordic replied:
Thanks!

kartben commented on doc/releases/release-notes-4.1.rst:
```diff
@@ -88,6 +88,17 @@ Deprecated APIs and options
+  1. when Stream Flash is not configured to do erase on its own
+  2. when erase is used for removal of a data prior or after Stream Flash uses the designated area.
+
+* Deprecated TinyCrypt library. The reasons for this are (:github:`43712`):
+
```
rebase glitch

```suggestion
```
